doc/designs/application-specific-passwords.md

```diff
@@ -0,0 +1,283 @@
+application-specific password for the respective application will be reset. A
+new parameter therefore needs to be added to the password reset command -
+application name (for the application-specific password).
+
+The user account status will display information for all its passwords.
+
+### ipa-kdb support for the aforementioned functionality
+
+The ipa-kdb daemon will be extended to support application-specific passwords
+as well, so that they can be used with Kerberos and HBAC.
+
+Subaccounts will be introduced with principals in the form 
+```_appname_/_user_@_REALM_```
+(to be consistent with other things in FreeIPA) - a username will remain the
+same for subaccounts, only the principal will contain the application name. The
+code of FreeIPA must be modified to take into account the newly changed
+assumptions. Until now, these were:
+* two-component principals with a host as the first component are always host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are always service principals
+  (`_service_/_fqdn_`)
+* single-component principals are always users (user)
+Now, they will be:
+* two-component principals with an application name as the first component are
+  users logged in with an application-specific password (`_appname_/_user_`)
+* two-component principals with a host as the first component are host
+  principals (`_host_/_fqdn_`)
+* all other two-component principals are service principals (`_service_/_fqdn_`)
+* single-component principals are users logged in with their primary password
+  (`user`)
+
+### Permissions management for application-specific passwords
+
+New ACIs must automatically be added and maintained for every
+application-specific password so that the owning user will have read-write
+access to all (and only) objects holding his application-specific passwords - 
+both when logged in using the primary password and when logged in using the
+respective application-specific password. This is the only thing to do with
+respect to RBAC and application-specific passwords, as RBAC serves only for work
+with LDAP in FreeIPA.
+
+The only general limitation for subaccounts will be the inability to manage the
+account (modify the information about the user, change his other passwords...).
+As all permissions for them will be assigned by administrators separately and
+from scratch, even a limitation such as reducing the effective permissions of
+the current user session (possibly according to the application policy) using a
+389-DS plugin probably will not be needed.
+
+#### HBAC for application-specific passwords
+
+Two-component principals for subaccounts, as described above, will be used for
+assigning HBAC rules to users logged in using application-specific passwords.
+These HBAC rules will be assigned by administrators just as the other ones.
+
+TBD: When an application-specific password is deleted, delete the associated
+HBAC rules automatically, or keep them?
```

Comment:
HBAC rules, as well need to be generic and catch any user's password, not a single user's app-password.
IE the HBAC rule should apply (normally) to either a group or a wildcard: */appname@REALM where '*' means any user.
One way to handle this is also to use a rule for automatic group creation via MEP plugin or similar, so that we do not have to introduce some new mechanism just for this, in that case we'll have a group entry named after the appname and the HBAC rule applying to that group.
When a user creates a new app password for that app the entry will be automatically added to the group.
The only issue to balance is group sizes. if we think groups may become too big then a wildcard based approach is better.

Reply:
This pair of proposals is an alternative to binding app-specific passwords to HBAC services and service groups that @tiran presented [above](https://github.com/freeipa/freeipa/pull/4061#discussion_r396405709).
We must analyze which way will be the most suitable. It seems to me that automatic group creation based on the app name (BTW, MEP stands for message exchange protocol?) would be the most difficult way to implement. Size of such groups might become an issue for heavily used applications, so using wildcarts might indeed be better. But these are just my preliminary impressions.

Reply:
MEP is 'Managed Entries Plugin' in 389-ds. You need to look at RHDS documentation for these details. https://access.redhat.com/documentation/en-us/red_hat_directory_server/10/html/deployment_guide/managed-entries

Reply:
We can actually use MEP w/o any group creation, just to handle the referential integrity.
As for HBAC rules, why do we need them again ?
HBAC for services is enforced locally on services by SSSD, is the plan to use HBAC somehow to control what password to be used by SSSD for a specific service on a specific machine?
Not entirely sure HBAC is the right tool, as you may have an app password for an "app" and yet have different HBAC rules for actual access to various servers.

Reply:
I've had a look at MEP, and it seems more promising that I previously thought - but of course, group size still may be a problem. Now I apologize, but I don't understand what referential integrity would be kept be MEP, if no groups would be created.

Regarding HBAC, it is a second proposal to handle an ambition stated in the issue 4510, which formed the assignment of my thesis:

> Ideally, this model could be quite sophisticated, and allow you to configure access based on the password used to log into an account - if I log in with the 'master' password (or, even better, the master 2FA token!) I get full access to the account, including changing all the other passwords and access to all services, but if I log in with an application specific password, I can only access a particular subset of systems and services that password is allowed access to (e.g., it could grant access only to log in to the mail server).

So as for your question

> is the plan to use HBAC somehow to control what password to be used by SSSD for a specific service on a specific machine?

Yes, that's it. There is.
Originally, I envisaged some new user-administered mechanism of permission management for app-specific passwords, with some constraints set by administrators, and some sort of a 'gearing mechanism' bringing this all together. This would be complicated, and after the first round of discussions, we abandoned this idea.
HBAC applied to subaccounts with appname in the principal was another proposed solution. However, it is getting clear that at least HBAC rules will not be created and applied to every single subaccount separately, but there will be some grouping, probably application-based.

As for the last problem you have outlined:

> you may have an app password for an "app" and yet have different HBAC rules for actual access to various servers

This is a good point. The most primitive solution to this is having one HBAC rule for the given app for each server it can be accessed on. A better solution would be creating a hostgoup for the given app, and creating a HBAC rule for this hostgroup.

I don't know what other mechanism than HBAC could we use for the aforementioned ambition.

Reply:
If we create a mapping mechanism between app name and HBAC service then we do not really need to mess with groups. Just add a new service for HBAC rules to use.
Hostgroups then will deal with different policies for different servers for the same "app".

Reply:
What is tricky is to define when it is ok to let a system use any app passwords.

We'd need a wildcard mechanism that says: this service can be access with any app password created by users, just like in Google's 2FA you can create your own app password and then use it for any random service (be it calendar, or mail, or...)

Reply:
> If we create a mapping mechanism between app name and HBAC service then we do not really need to mess with groups. Just add a new service for HBAC rules to use.

This proposal is partially similar to the one presented [above](https://github.com/freeipa/freeipa/pull/4061#discussion_r396405709) by @tiran. But unlike that proposal, the user would be free to set any application name; however, they would still need to be mapped to already existing or newly created HBAC services or service groups. But it is clear something new must be created for this feature, and this way may be the best - I see the proposed mapping mechanism as at least "cleaner", and possibly also simpler, than creating new groups using a MEP plugin.

> Hostgroups then will deal with different policies for different servers for the same "app".

All right, apparently there is no better way to do it.

> What is tricky is to define when it is ok to let a system use any app passwords.
> We'd need a wildcard mechanism that says: this service can be access with any app password created by users, just like in Google's 2FA you can create your own app password and then use it for any random service (be it calendar, or mail, or...)

All right, so a wildcart will be used to specify 'all the user's app-specific passwords'. What character will we use as a wildcart - an asterisk (`*`)?